### ISSUE TITLE
MQTT Service reconnects every time the `publish` method is called

### ISSUE THREAD

**Issue by PhilippMatthes** [FraunhoferIOSB/FROST-Client#120]
## Problem description

We use a setup as follows

```java
final MqttConnectOptions mqttConnectOptions = new MqttConnectOptions();
mqttConnectOptions.setCleanSession(true);
mqttConnectOptions.setConnectionTimeout(0);
// ...
final SensorThingsService service = new SensorThingsService(serviceURL, mqttConfig);
// ...
final Observation observation = new Observation();
// ...
service.publish(topic, observation);
```

In our use case, we want to publish multiple observations to a Mosquitto broker in a short period of time. When we do this, we observe two problems:
1. The Mosquitto broker will log a connect - publish - disconnect for each published message
2. The host application crashes when subsequent calls are in a short period of time with `de.fraunhofer.iosb.ilt.sta.MqttException: Error publishing via MQTT` - `Caused by: No connection to Client`.

However, from our POV, the MQTT client should keep the connection open to allow for fast subsequent publish calls.

## Workaround

We managed to workaround this problem by using Paho MQTT directly, as follows:

```java
// Generate a unique client id
final String clientId = "java-client-" + UUID.randomUUID().toString();
client = new MqttClient(mqttEndpoint, clientId);
client.connect(mqttConnectOptions);
// ...
client.publish(topic, new MqttMessage(ObjectMapperFactory.get().writeValueAsBytes(observation)));
// ...
client.disconnect();
```  

## Possible Solution

We could not locate why the `SensorThingsService` disconnects after a successful `publish` call. We recommend to investigate the problem using our workaround given above.

**Comment by hylkevds** [comment]
Are there any differences in the configuration you give Paho in both cases?
Are you setting the clientId when using FROST-Client?

**Comment by PhilippMatthes** [comment]
There are no differences in configuration, other than the client id which we set manually as seen above. With the `SensorThingsService`, we use the default client id which we figured is generated in a similar manner as seen above.

**Comment by hylkevds** [comment]
Something has to be different, or the problem would not exist.
Can you make a minimal-test-case that demonstrates the problem and the workaround?

**Comment by PhilippMatthes** [comment]
Of course. We use the following docker compose setup to host our Service.

<details>

```
version: '2'

services:
  web:
    image: fraunhoferiosb/frost-server-http:1.14.0
    environment:
      - serviceRootUrl=http://localhost:8080/FROST-Server
      - http_cors_enable=true
      - http_cors_allowed.origins=*
      - persistence_db_driver=org.postgresql.Driver
      - persistence_db_url=jdbc:postgresql://database:5432/sensorthings
      - persistence_db_username=sensorthings
      - persistence_db_password=ChangeMe
      - persistence_autoUpdateDatabase=true
    ports:
      - 20055:8080
    depends_on:
      - database

  mqtt:
    image: eclipse-mosquitto:2.0.14
    hostname: mosquitto
    logging:
      options:
        # Rotate logging files after 512k
        # See: https://docs.docker.com/compose/compose-file/compose-file-v3/
        max-size: '512k'
    ports:
      - 20056:1883
    volumes:
      # Mandatory volumes for the mosquitto MQTT broker
      - ./mosquitto/config/:/mosquitto/config/
    restart: unless-stopped

  database:
    image: postgis/postgis:11-2.5-alpine
    environment:
      - POSTGRES_DB=sensorthings
      - POSTGRES_USER=sensorthings
      - POSTGRES_PASSWORD=ChangeMe
    volumes:
      - postgis_volume:/var/lib/postgresql/data
volumes:
    postgis_volume:

```

</details>

With the following Mosquitto config:

<details>

```
listener 1883
protocol mqtt
allow_anonymous true
max_inflight_messages 0
max_queued_messages 0
```

</details>

Finally, here is the code to reproduce the issue:

<details>

```java
package example;

import java.net.MalformedURLException;
import java.net.URL;
import java.time.ZonedDateTime;

import org.eclipse.paho.client.mqttv3.MqttConnectOptions;

import de.fraunhofer.iosb.ilt.sta.MqttException;
import de.fraunhofer.iosb.ilt.sta.model.Observation;
import de.fraunhofer.iosb.ilt.sta.service.MqttConfig;
import de.fraunhofer.iosb.ilt.sta.service.SensorThingsService;

public class App {
    private static final String mqttEndpoint = "tcp://localhost:20056";
    private static final String serviceEndpoint = "http://localhost:20055/FROST-Server/v1.1/";

    private static SensorThingsService service;
    
    public static void main(String[] args) throws InterruptedException {
        // Initialize the service with the service endpoint.
        final MqttConnectOptions mqttConnectOptions = new MqttConnectOptions();
        mqttConnectOptions.setCleanSession(true);
        mqttConnectOptions.setConnectionTimeout(0);

        final MqttConfig config = new MqttConfig(mqttEndpoint);
        config.setOptions(mqttConnectOptions);

        try {
            final URL serviceEndpointURL = new URL(serviceEndpoint);
            service = new SensorThingsService(serviceEndpointURL, config);
        } catch (MalformedURLException e) {
            throw new RuntimeException("Malformed service endpoint URL", e);
        } catch (MqttException e) {
            throw new RuntimeException("Could not connect to MQTT broker", e);
        }

        for (int i = 0; i < 10; i++) {
            Observation observation = new Observation();
            observation.setPhenomenonTimeFrom(ZonedDateTime.now());
            observation.setResult(i);
            observation.setResultTime(ZonedDateTime.now());

            final String topic = "observations";
            try {
                service.publish(topic, observation);
                System.out.println("Published observation " + i + " to topic " + topic);
            } catch (MqttException e) {
                throw new RuntimeException("Could not publish observation", e);
            }
        }
    }
}
```

</details>

With the running docker-compose setup, the Java app will produce something like:

```
Published observation 0 to topic observations
Published observation 1 to topic observations
Exception in thread "main" java.lang.RuntimeException: Could not publish observation
        ...
```

**Comment by PhilippMatthes** [comment]
You should be able to apply the workaround by the initial problem description.

**Comment by hylkevds** [comment]
Hmm, You're not using the STA MQTT API at all, you're simply using a standard broker. This will not result in Observations being created in the database...
If you want to use MQTT to create Observations in your database, or to subscribe to changes in Entities, you have to use the FROST-MQTT component.

**Comment by PhilippMatthes** [comment]
I can confirm that we don't use the MQTT component of the FROST Server, this is on purpose and has reasons within our system, that are irrelevant for this issue. I have updated the docker-compose setup in my comment, so that it is easier to follow along for this issue. As initially described, we use Mosquitto to publish the observations.

**Comment by hylkevds** [comment]
It seems you're the first trying to publish to MQTT using FROST-Client... the deliveryComplete callback would always throw an UnsupportedOperationException, causing the disconnect. 
Fixed in 0.41.

**Comment by PhilippMatthes** [comment]
Thanks, feel free to close this issue when the problem was fixed.
